### PR TITLE
Updating error handling for compound retrievers

### PR DESCRIPTION
**Pull request by pmpailis**
Updating error handling for compound retrievers so that if any of the requests raises a `5xx` error, then this is propagated to the response, otherwise, it we throw a `4xx` instead. All nested failures are added as suppressed ones to the error returned to the user.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/retriever/CompoundRetrieverBuilder.java [hunk 1]
```diff
@@ -134,9 +138,25 @@ public void onResponse(MultiSearchResponse items) {
...
                     }
                     if (false == failures.isEmpty()) {
-                        IllegalStateException ex = new IllegalStateException("Search failed - some nested retrievers returned errors.");
+                        boolean serverErrorFound = false;
+                        for (Exception failure : failures) {
+                            if (ExceptionsHelper.status(failure).getStatus() >= RestStatus.INTERNAL_SERVER_ERROR.getStatus()) {
+                                serverErrorFound = true;
+                                break;
+                            }
+                        }
```
[the remark this answers is not in the document]
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/retriever/CompoundRetrieverBuilder.java]: Updating in 072f5b9bb7ba451d08e3f2c6b0d9c36e83128ae8

**Comment by pmpailis** [comment]
@elasticmachine update branch

**Comment by pmpailis** [comment]
@elasticmachine update branch

File: x-pack/plugin/rank-rrf/src/internalClusterTest/java/org/elasticsearch/xpack/rank/rrf/RRFRetrieverBuilderIT.java [hunk 2]
```diff
@@ -615,9 +618,53 @@ public void testRRFInnerRetrieverSearchError() {
...
+        Exception ex = expectThrows(ElasticsearchStatusException.class, req::get);
+        assertThat(ex, instanceOf(ElasticsearchStatusException.class));
+        assertThat(
+            ex.getMessage(),
+            containsString(
+                "[rrf] search failed - retrievers '[standard]' returned errors. All failures are attached as suppressed exceptions."
+            )
+        );
+        assertThat(ExceptionsHelper.status(ex), equalTo(RestStatus.BAD_REQUEST));
+        assertThat(ex.getSuppressed().length, greaterThan(0));
```
Comment (carlosdelest) [inline x-pack/plugin/rank-rrf/src/internalClusterTest/java/org/elasticsearch/xpack/rank/rrf/RRFRetrieverBuilderIT.java:630]: We know the number of suppressed exceptions we should have, right? Can we check here?

**Review by carlosdelest — APPROVED** [review]
This LGTM 👍

File: server/src/main/java/org/elasticsearch/search/retriever/CompoundRetrieverBuilder.java [hunk 3]
```diff
@@ -122,10 +124,12 @@ public final RetrieverBuilder rewrite(QueryRewriteContext ctx) throws IOExceptio
                 public void onResponse(MultiSearchResponse items) {
                     List<ScoreDoc[]> topDocs = new ArrayList<>();
                     List<Exception> failures = new ArrayList<>();
+                    List<String> retrieversWithFailures = new ArrayList<>();
                     for (int i = 0; i < items.getResponses().length; i++) {
                         var item = items.getResponses()[i];
                         if (item.isFailure()) {
                             failures.add(item.getFailure());
+                            retrieversWithFailures.add(innerRetrievers.get(i).retriever().getName());
```
[the remark this answers is not in the document]
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/retriever/CompoundRetrieverBuilder.java:134]: Updated in 1e7c5b38e2f07064d53b846916cdb0920d8afa60

File: x-pack/plugin/rank-rrf/src/internalClusterTest/java/org/elasticsearch/xpack/rank/rrf/RRFRetrieverBuilderIT.java (hunk 2 above)
Comment (pmpailis) [reply x-pack/plugin/rank-rrf/src/internalClusterTest/java/org/elasticsearch/xpack/rank/rrf/RRFRetrieverBuilderIT.java]: Yeap, makes sense! Updated in 1e7c5b38e2f07064d53b846916cdb0920d8afa60

**Comment by pmpailis** [comment]
@elasticmachine update branch

**Pull request merged**